File: demo/so-bindings/julia/ffi_example.jl
lib = Sys.isapple() ? "example.dylib" : "example.so"

for n in 1:10
    result = ccall((:factorial, lib), Clonglong, (Clonglong,), n)
    println("$(n)! = $(result)")
end

r = ccall((:greet, lib), Cstring, (Cstring,), "World")
println(unsafe_string(r))

r = ccall((:repeat_string, lib), Cstring, (Cstring, Clonglong), "ha", 3)
println(unsafe_string(r))

ccall((:shout_it, lib), Cvoid, (Cstring,), "hello from yamlscript")

m = ccall((:maybe, lib), Cint, ())
println("maybe: $(m != 0)")

r = ccall((:sort_json_array, lib), Cstring, (Cstring,), "[3,1,4,1,5,9,2,6]")
println("sorted: $(unsafe_string(r))")

File: demo/so-bindings/nim/ffi_example.nim
when defined(macosx):
  {.passL: "-lexample".}
else:
  {.passL: "-l:example.so".}

proc factorial(n: clonglong): clonglong {.importc, cdecl.}
proc greet(name: cstring): cstring {.importc, cdecl.}
proc repeat_string(s: cstring; n: clonglong): cstring {.importc, cdecl.}
proc shout_it(s: cstring) {.importc, cdecl.}
proc maybe(): cint {.importc, cdecl.}
proc sort_json_array(s: cstring): cstring {.importc, cdecl.}

for n in 1..10:
  echo $n & "! = " & $factorial(n.clonglong)

echo $greet("World")
echo $repeat_string("ha", 3)

shout_it("hello from yamlscript")

let m = maybe()
echo "maybe: " & (if m != 0: "true" else: "false")

echo "sorted: " & $sort_json_array("[3,1,4,1,5,9,2,6]")
